### PR TITLE
KAFKA-17285: Consider using `Utils.closeQuietly` to replace `CoreUtils.swallow` when handling Closeable objects

### PR DESCRIPTION
**Pull request by bboyleonp666**
Use `Utils.closeQuietly` to replace `CoreUtils.swallow` in `kafka.network.Acceptor#closeAll`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by bboyleonp666** [comment]
Using `Utils.closeQuietly` to replace `CoreUtils.swallow` will also change the logging level from **ERROR** to **WARNING**, which makes sense to me, since the `close()` called before the thread starting is not supposed to happen. The warning message is enough for this behavior.

**Comment by chia7712** [comment]
@bboyleonp666 could you please grep code base to fix all similar issues?

**Comment by bboyleonp666** [comment]
Sure. Let me take a look at the code base.

**Comment by bboyleonp666** [comment]
@chia7712 While handling this, I have noticed that there are lots of methods that uses `shutdown()` instead of `close()` when calling `CoreUtils.swallow()`. Currently, only `close()` is handled by `Utils.closeQuietly()` and I cannot find a method that can gracefully deal with `shutdown()` method. Are there any methods that can do this? If not, I wonder if we are going to add one similarly. Maybe name it `shutdownQuietly`?

**Comment by chia7712** [comment]
> Are there any methods that can do this? If not, I wonder if we are going to add one similarly. Maybe name it shutdownQuietly?

the `shutdown` is not a kind of interface, so I guess it would be hard to add "one helper" to handle "all" cases. Maybe this PR can focus on deal with the `AutoClosable` object.

**Comment by bboyleonp666** [comment]
I agree with you. I have already merged the commits in trunk.

**Comment by chia7712** [comment]
@bboyleonp666 please rebase code to fix build error. Also, please fix following usages:

https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala#L72

https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/main/scala/kafka/admin/AclCommand.scala#L221

https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/main/scala/kafka/network/SocketServer.scala#L420

https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/main/scala/kafka/raft/KafkaMetadataLog.scala#L684

https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/main/scala/kafka/server/ControllerServer.scala#L70

https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/main/scala/kafka/server/ZkAdminManager.scala#L550

https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/main/scala/kafka/server/ZkAdminManager.scala#L551

**Comment by chia7712** [comment]
@bboyleonp666 any updates?

**Comment by bboyleonp666** [comment]
I was on a long trip just back to work recently. I will update this soon, thanks for reminding.

File: core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala [hunk 1]
```diff
@@ -18,7 +18,7 @@
 package kafka.zk
 
 import org.apache.zookeeper.server.{NIOServerCnxnFactory, SessionTrackerImpl, ZooKeeperServer}
-import kafka.utils.{CoreUtils, Logging, TestUtils}
+import kafka.utils.{Exit, Logging, TestUtils}
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala]: Please revert this change :)

**Review by chia7712 — COMMENTED** [review]
@bboyleonp666 thanks for your updates

File: core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala (hunk 1 above)
Comment (bboyleonp666) [reply core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala]: Oh! Thanks. didn't notice this one.

**Comment by bboyleonp666** [comment]
> @bboyleonp666 please rebase code to fix build error. Also, please fix following usages:
> 
> https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala#L72

Since `zookeeper.getZKDatabase` is not a `Closeable`, we are not able to use `Utils.closeQuietly()` to replace it. I will keep it as it is.

> 
> https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/main/scala/kafka/server/ZkAdminManager.scala#L550
> 
> https://github.com/apache/kafka/blob/28cd7136fb1b5702af82c3736d246c7a1e709140/core/src/main/scala/kafka/server/ZkAdminManager.scala#L551

The lines here use `foreach` to close the resources, while `CoreUtils.swallow()` is not limited to an object type like `Utils.closeQuietly()` does. I wonder whether we should keep using `foreach` to keep the code concise and consistent or should we make it a for loop to achieve the same effect? What's your opinion? @chia7712

**Comment by chia7712** [comment]
> The lines here use foreach to close the resources, while CoreUtils.swallow() is not limited to an object type like Utils.closeQuietly() does. I wonder whether we should keep using foreach to keep the code concise and consistent or should we make it a for loop to achieve the same effect? What's your opinion? @chia7712

how about: `createTopicPolicy.foreach(Utils.closeQuietly(_, "close create topic policy"))`

**Comment by bboyleonp666** [comment]
Looks great. Let me apply this change to all the instances that I can find.

File: core/src/main/scala/kafka/server/ControllerServer.scala [hunk 2]
```diff
@@ -516,9 +512,9 @@ class ControllerServer(
         controller.close()
       if (quorumControllerMetrics != null)
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/ControllerServer.scala]: Could you please apply to `quorumControllerMetrics`, `controller` and other closable variables?

File: core/src/main/scala/kafka/server/SharedServer.scala [hunk 3]
```diff
@@ -366,34 +366,22 @@ class SharedServer(
...
-      if (snapshotGenerator != null) {
-        CoreUtils.swallow(snapshotGenerator.close(), this)
-        snapshotGenerator = null
-      }
+      Utils.closeQuietly(loader, "loader")
+      loader = null
+      Utils.closeQuietly(metadataLoaderMetrics, "metadata loader metrics")
+      metadataLoaderMetrics = null
+      Utils.closeQuietly(snapshotGenerator, "snapshot generator")
+      snapshotGenerator = null
```
Comment (bboyleonp666) [inline core/src/main/scala/kafka/server/SharedServer.scala:373]: Hey @chia7712, I have no idea why these objects still being assigned null value after close. Thus, I keep this behavior to all the changes for now. Should I keep them or remove them?

File: core/src/main/scala/kafka/server/SharedServer.scala [hunk 4]
```diff
@@ -366,34 +366,22 @@ class SharedServer(
...
-      if (metrics != null) {
-        CoreUtils.swallow(metrics.close(), this)
-        metrics = null
-      }
+      Utils.closeQuietly(controllerServerMetrics, "controller server metrics")
+      controllerServerMetrics = null
+      Utils.closeQuietly(brokerMetrics, "broker metrics")
+      brokerMetrics = null
+      Utils.closeQuietly(metrics, "metrics")
+      metrics = null
```
Comment (bboyleonp666) [inline core/src/main/scala/kafka/server/SharedServer.scala:383]: ditto

File: core/src/main/scala/kafka/server/ControllerServer.scala [hunk 5]
```diff
@@ -487,18 +481,12 @@ class ControllerServer(
...
-      if (registrationsPublisher != null) {
-        registrationsPublisher.close()
-        registrationsPublisher = null
-      }
+      Utils.closeQuietly(metadataCachePublisher, "metadata cache publisher")
+      metadataCachePublisher = null
+      Utils.closeQuietly(featuresPublisher, "features publisher")
+      featuresPublisher = null
+      Utils.closeQuietly(registrationsPublisher, "registrations publisher")
+      registrationsPublisher = null
```
Comment (bboyleonp666) [inline core/src/main/scala/kafka/server/ControllerServer.scala:490]: ditto

File: core/src/main/scala/kafka/server/SharedServer.scala (hunk 3 above)
Comment (chia7712) [reply core/src/main/scala/kafka/server/SharedServer.scala:373]: > Should I keep them or remove them?

please keep them :)

Comment (bboyleonp666) [reply core/src/main/scala/kafka/server/SharedServer.scala:373]: Ok!

File: core/src/main/scala/kafka/server/BrokerServer.scala [hunk 6]
```diff
@@ -756,7 +756,7 @@ class BrokerServer(
 
       // Close remote log manager to give a chance to any of its underlying clients
       // (especially in RemoteStorageManager and RemoteLogMetadataManager) to close gracefully.
-      CoreUtils.swallow(remoteLogManagerOpt.foreach(_.close()), this)
+      remoteLogManagerOpt.foreach(Utils.closeQuietly(_, "remote log manager option"))
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/BrokerServer.scala]: `remote log manager`

File: core/src/main/scala/kafka/server/ControllerServer.scala [hunk 7]
```diff
@@ -512,13 +500,11 @@ class ControllerServer(
...
-        controller.close()
-      if (quorumControllerMetrics != null)
-        CoreUtils.swallow(quorumControllerMetrics.close(), this)
-      CoreUtils.swallow(authorizer.foreach(_.close()), this)
-      createTopicPolicy.foreach(policy => CoreUtils.swallow(policy.close(), this))
-      alterConfigPolicy.foreach(policy => CoreUtils.swallow(policy.close(), this))
+      Utils.closeQuietly(controller, "controller")
+      Utils.closeQuietly(quorumControllerMetrics, "quorum controller metrics")
+      authorizer.foreach(Utils.closeQuietly(_, "authorizer"))
+      createTopicPolicy.foreach(policy => Utils.closeQuietly(policy, "topic policy"))
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/ControllerServer.scala]: `create topic policy`

File: core/src/main/scala/kafka/server/KafkaServer.scala [hunk 8]
```diff
@@ -1051,13 +1051,12 @@ class KafkaServer(
         // Close remote log manager before stopping processing requests, to give a chance to any
         // of its underlying clients (especially in RemoteStorageManager and RemoteLogMetadataManager)
         // to close gracefully.
-        CoreUtils.swallow(remoteLogManagerOpt.foreach(_.close()), this)
+        remoteLogManagerOpt.foreach(Utils.closeQuietly(_, "remote log manager option"))
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaServer.scala]: `remote log manager`

File: core/src/main/scala/kafka/server/KafkaServer.scala [hunk 9]
```diff
@@ -1068,8 +1067,7 @@ class KafkaServer(
         if (socketServer != null)
           CoreUtils.swallow(socketServer.shutdown(), this)
         unregisterCurrentControllerIdMetric()
-        if (metrics != null)
-          CoreUtils.swallow(metrics.close(), this)
+        Utils.closeQuietly(metrics, "metrics")
         if (brokerTopicStats != null)
           CoreUtils.swallow(brokerTopicStats.close(), this)
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaServer.scala]: Could you please make `brokerTopicStats` extend `Closeable`?

File: core/src/main/scala/kafka/server/ZkAdminManager.scala [hunk 10]
```diff
@@ -547,8 +547,8 @@ class ZkAdminManager(val config: KafkaConfig,
 
   def shutdown(): Unit = {
     topicPurgatory.shutdown()
-    CoreUtils.swallow(createTopicPolicy.foreach(_.close()), this)
-    CoreUtils.swallow(alterConfigPolicy.foreach(_.close()), this)
+    createTopicPolicy.foreach(Utils.closeQuietly(_, "topic policy"))
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/ZkAdminManager.scala]: `create topic policy`

**Review by chia7712 — COMMENTED** [review]
@bboyleonp666 thanks for updated patch

File: core/src/main/scala/kafka/server/KafkaServer.scala (hunk 9 above)
Comment (bboyleonp666) [reply core/src/main/scala/kafka/server/KafkaServer.scala]: Sounds great! Add it in 21f9b18.

**Comment by chia7712** [comment]
@bboyleonp666 could you please fix conflicts?

**Comment by bboyleonp666** [comment]
Hi @chia7712, thank you. Just resolved and waiting for the checks.

File: core/src/main/scala/kafka/server/BrokerServer.scala [hunk 11]
```diff
@@ -766,14 +766,13 @@ class BrokerServer(
         CoreUtils.swallow(socketServer.shutdown(), this)
       if (brokerTopicStats != null)
         CoreUtils.swallow(brokerTopicStats.close(), this)
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/BrokerServer.scala]: `BrokerTopicStats` extends `Closeable` now, so you can apply `Utils.closeQuietly`, right?

Comment (bboyleonp666) [reply core/src/main/scala/kafka/server/BrokerServer.scala]: Exactly. Let me update this one.

Comment (bboyleonp666) [reply core/src/main/scala/kafka/server/BrokerServer.scala]: Apply this one in commit fd93c6b

**Review by chia7712 — APPROVED** [review]
